src/Data/Deque/Mutable/Linear.hs

```diff
@@ -0,0 +1,182 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+-- | Mutable Linear Deque
+--
+-- This module provides a pure interface to a mutable deque.
+--
+-- It is designed to be imported qualfied:
+--
+-- > import qualfied Data.Deque.Mutable.Linear as Deque
```

Comment:
I think it is crucial to mention that this is a **bounded** deque.  I would even call it `Data.Dequeue.Bounded.Mutable.Linear` instad. It would also be good to mention:

* That the underlying storage is an array
* When full, newer elements overwrite the older elements

Reply:
It doesn't overwrite. It crashes if you try to fill it after it's full. I think that makes more sense.

Reply:
I am having a hard time thinking when would I require a queue where I certainly know that it will contain less than some number of elements. I can think of cases where it's okay to be forgetful, or don't accept the new element, or in concurrent cases (which is not the case in this code) it should block; but in all cases I can think of there is no guarantee that the queue is not empty or it is not full. 

Even if we go with the crashy version, we should still mention that it is a bounded queue.

Reply:
I agree with adding comments saying it is bounded. But ...

Do you think it's just better for me to use array resizing internally and have an unbounded structure to begin with? I feel like your argument makes sense and I can't think of a case where we want a bounded queue or stack.

Reply:
Did this ^, done.